balancer_conn_wrappers.go

```diff
@@ -148,42 +172,144 @@ func (ccb *ccBalancerWrapper) resolverError(err error) {
+func (ccb *ccBalancerWrapper) enterIdleMode() {
+	channelz.Info(logger, ccb.cc.channelzID, "ccBalancerWrapper: entering idle mode")
+	ccb.handleCloseAndEnterIdle(ccbModeIdle)
+}
+
+// handleCloseAndEnterIdle is invoked when the channel is being closed or when
+// it enters idle mode upon expiry of idle_timeout.
+//
+// This call is not scheduled on the serializer because we need to ensure that
+// the current serializer is completely shutdown before the next one is created
+// (when exiting idle).
+func (ccb *ccBalancerWrapper) handleCloseAndEnterIdle(m ccbMode) {
+	ccb.mu.Lock()
+	if ccb.mode == ccbModeClosed || ccb.mode == ccbModeIdle {
+		ccb.mu.Unlock()
+		return
+	}
+
+	// Close the serializer to ensure that no more calls from gRPC are sent
+	// to the balancer.
+	ccb.serializerCancel()
+	ccb.mode = m
+	done := ccb.serializer.Done
+	b := ccb.balancer
+	ccb.mu.Unlock()
+
+	// Give enqueued callbacks a chance to finish.
+	<-done
+	// Spawn a goroutine to close the balancer (since it may block trying to
+	// cleanup all allocated resources) and return early.
+	go b.Close()
+}
+
+// exitIdleMode is invoked by grpc when the channel exits idle mode either
+// because of an RPC or because of an invocation of the Connect() API. This
+// recreates the balancer that was closed previously when entering idle mode.
+//
+// If the channel is not in idle mode, we know for a fact that we are here as a
+// result of the user calling the Connect() method on the ClientConn. In this
+// case, we can simply forward the call to the underlying balancer, instructing
+// it to reconnect to the backends.
+func (ccb *ccBalancerWrapper) exitIdleMode() {
+	ccb.mu.Lock()
+	if ccb.mode == ccbModeClosed {
+		// Request to exit idle is a no-op when wrapper is already closed.
+		ccb.mu.Unlock()
+		return
+	}
+
+	if ccb.mode == ccbModeIdle {
+		// Recreate the serializer which was closed when we entered idle.
+		ctx, cancel := context.WithCancel(context.Background())
+		ccb.serializer = grpcsync.NewCallbackSerializer(ctx)
+		ccb.serializerCancel = cancel
+	}
 
-		if err := ccb.balancer.SwitchTo(builder); err != nil {
-			channelz.Errorf(logger, ccb.cc.channelzID, "Channel failed to build new LB policy %q: %v", name, err)
+	// The ClientConn guarantees that mutual exclusion between close() and
+	// exitIdleMode(), and since we just created a new serializer, we can be
+	// sure that the below function will be scheduled.
+	done := make(chan struct{})
+	ccb.serializer.Schedule(func(_ context.Context) {
+		defer close(done)
+
+		ccb.mu.Lock()
+		defer ccb.mu.Unlock()
+
+		if ccb.mode != ccbModeIdle {
+			ccb.balancer.ExitIdle()
 			return
 		}
-		ccb.curBalancerName = builder.Name()
+
+		// Gracefulswitch balancer does not support a switchTo operation after
+		// being closed. Hence we need to create a new one here.
+		ccb.balancer = gracefulswitch.NewBalancer(ccb, ccb.opts)
+		ccb.buildLoadBalancingPolicy(ccb.curBalancerName)
```

Comment:
Is `ccb.curBalancerName` valid at this point?  When we are exiting idle, I assume we haven't heard from the new name resolver yet, still, right?  So we should wait until we get an update from it and do this exactly the same way we initially create the `ccbWrapper`.

Ideally...

Can we make `ClientConn.Dial` always start in the idle state and then call `cc.exitIdleMode()`?  I think that would eliminate some initialization code there that is effectively duplicated by the support for exiting idle mode.

Reply:
Probably the code just isn't ready to review yet, but I just wanted to ping this because I think it's important, and you did respond to several other comments.

Reply:
I handled this in a more recent commit by resetting it to the empty string.